### PR TITLE
CDAP-3657: Improve Id serialization

### PR DESCRIPTION
**Pull request by jwang47**
https://issues.cask.co/browse/CDAP-3657
http://builds.cask.co/browse/CDAP-RBT423

### CODE REVIEW

**Comment by bdmogal** [comment]
@jwang47 not reviewed the PR, but we should test upgrade thoroughly before merging this. Will this impact serialization of any ``Id`` objects in CDAP meta tables - ``app.meta``, ``usage.registry``, ``dataset.instance``, ``dataset.type``, etc?

**Comment by poornachandra** [comment]
It might be a good idea to add a comment on top of `toString()` methods saying that they are part of external contract, and cannot be changed without checking for backwards compatibility.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java [hunk 1]
```diff
@@ -59,33 +53,19 @@ private static boolean isValidDatasetId(String datasetId) {
...
-  public String toString() {
-    return getIdRep();
+  public String getIdType() {
+    return getIdType(getClass());
   }
 
-  @Nullable
-  protected abstract Id getParent();
-
-  public abstract String getId();
```
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Any reason for the removal of this method? I should be able to get an Id from an Id object, right? This breaks compilation of LineageRecord class.

Comment (jwang47) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: This method doesn't make sense for all `Id`s, so I removed it. However, I can simply make `toString()` return e.g. `program:namespace1.app1.flow.flow1` and `getId()` return `namespace1.app1.flow.flow1`

**Comment by jwang47** [comment]
@bdmogal The only places we're using `Id.toString()` is in `DatasetServiceClient`/`DatasetInstanceHandler` (internal API) and lineage, so there won't be any upgrade step needed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 2]
```diff
@@ -482,7 +482,7 @@ WorkflowStatistics getWorkflowStatistics(Id.Workflow workflowId, long startTime,
    * @param timeInterval The timeInterval with which the user wants to space out the runs
    * @return Map of runId of Workflow to DetailedStatistics of the run
    */
-  Collection<WorkflowDataset.WorkflowRunRecord> retrieveSpacedRecords(Id.Workflow workflow, String runId,
+  Collection<WorkflowDataset.WorkflowRunRecord> retrieveSpacedRecords(Id.Program workflow, String runId,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: why do we need these changes?

Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: Only `Id.Program` will be used from now on, rather than both `Id.Program` and `Id.Workflow`, etc. `Id.Workflow` will no longer be instantiable and will only contain helper methods.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/ConsumerSupplier.java [hunk 3]
```diff
@@ -47,24 +46,27 @@
   private final int numGroups;
   private final UsageRegistry usageRegistry;
   private final Id.Namespace namespace;
-  private final List<Id> owners;
+  private final Iterable<? extends Id.NamespacedId> owners;
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/ConsumerSupplier.java]: why does the list change to Iterable here?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java [hunk 4]
```diff
@@ -392,8 +392,10 @@ public void testBatchStatus() throws Exception {
     Assert.assertEquals(new NotFoundException(Id.Program.from("testnamespace1", "WordCountApp", ProgramType.FLOW,
                                                               "NotExist")).getMessage(),
                         returnedBody.get(0).get("error").getAsString());
-    Assert.assertEquals("'namespace:testnamespace1/application:WordCountApp/program:flows:NotExist' was not found.",
-                        returnedBody.get(0).get("error").getAsString());
+    // namespace:testnamespace1/application:WordCountApp/program:flows:NotExist
+    Assert.assertEquals(
+      String.format("'%s' was not found.", Id.Flow.from(TEST_NAMESPACE1, "WordCountApp", "NotExist").toString()),
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java]: can just use new ``new NotFoundException(Id.Flow.from(TEST_NAMESPACE1, "WordCountApp", "NotExist")).getMessage()`` here, so the message becomes internal to the Exception class?

Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java]: Please change other instances in this file that have a similar pattern

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java [hunk 5]
```diff
@@ -84,16 +89,18 @@ public EndpointStrategy get() {
   }
 
   @Nullable
-  public DatasetMeta getInstance(String instanceName, @Nullable Iterable<? extends Id> owners)
-    throws DatasetManagementException {
+  public DatasetMeta getInstance(String instanceName,
+                                 @Nullable Iterable<? extends Id> owners)
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: this can also be ``Iterable<? extends Id.NamespacedId>``?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java [hunk 6]
```diff
@@ -199,7 +200,8 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
   public <T extends Dataset> T getDataset(
     Id.DatasetInstance datasetInstanceId, Map<String, String> arguments,
     @Nullable ClassLoader classLoader,
-    @Nullable Iterable<? extends Id> owners) throws DatasetManagementException, IOException {
+    @Nullable Iterable<? extends Id> owners)
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: ``Iterable<? extends Id.NamespacedId>``

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java [hunk 7]
```diff
@@ -199,7 +200,8 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
...
+    throws DatasetManagementException, IOException, ServiceUnavailableException {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: ``ServiceUnavailableException`` is unchecked. It just got removed in another PR. You may have conflicts.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java [hunk 8]
```diff
@@ -218,7 +220,8 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
     Id.DatasetInstance datasetInstanceId, @Nullable Map<String, String> arguments,
     ClassLoader classLoader,
     DatasetClassLoaderProvider classLoaderProvider,
-    @Nullable Iterable<? extends Id> owners) throws DatasetManagementException, IOException {
+    @Nullable Iterable<? extends Id> owners) throws DatasetManagementException, IOException,
+    ServiceUnavailableException {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: please remove ``ServiceUnavailableException`` if it exists after rebasing with release/3.2

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java [hunk 9]
```diff
@@ -218,7 +220,8 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
     Id.DatasetInstance datasetInstanceId, @Nullable Map<String, String> arguments,
     ClassLoader classLoader,
     DatasetClassLoaderProvider classLoaderProvider,
-    @Nullable Iterable<? extends Id> owners) throws DatasetManagementException, IOException {
+    @Nullable Iterable<? extends Id> owners) throws DatasetManagementException, IOException,
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: ``Iterable<? extends Id.NamespacedId>``

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java [hunk 10]
```diff
@@ -214,26 +214,11 @@ public void executeDataOp(HttpRequest request, HttpResponder responder, @PathPar
...
-        String ownerType = parts[0];
-        String ownerId = parts[1];
-        if (ownerType.equals(Id.getType(Id.Program.class))) {
-          return Id.Program.fromStrings(ownerId.split("/"));
-        } else {
-          return null;
-        }
-      }
-    });
+  private List<Id.NamespacedId> decodeOwners(String owners) throws UnsupportedEncodingException {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java]: Is owners ``@Nullable``? If so, mark it. If not, remove the check below.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DatasetFramework.java [hunk 11]
```diff
@@ -239,7 +239,7 @@ void updateInstance(Id.DatasetInstance datasetInstanceId, DatasetProperties prop
   @Nullable
   <T extends Dataset> T getDataset(Id.DatasetInstance datasetInstanceId, @Nullable Map<String, String> arguments,
                                    @Nullable ClassLoader classLoader, @Nullable Iterable<? extends Id> owners)
-    throws DatasetManagementException, IOException;
+    throws DatasetManagementException, IOException, ServiceUnavailableException;
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DatasetFramework.java]: remove ServiceUnavailableException

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DatasetFramework.java [hunk 12]
```diff
@@ -278,7 +278,7 @@ void updateInstance(Id.DatasetInstance datasetInstanceId, DatasetProperties prop
                                    @Nullable ClassLoader classLoader,
                                    DatasetClassLoaderProvider classLoaderProvider,
                                    @Nullable Iterable<? extends Id> owners)
-    throws DatasetManagementException, IOException;
+    throws DatasetManagementException, IOException, ServiceUnavailableException;
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DatasetFramework.java]: ditto

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java [hunk 13]
```diff
@@ -159,7 +160,7 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
   public <T extends Dataset> T getDataset(Id.DatasetInstance datasetInstanceId,
                                           @Nullable Map<String, String> arguments, @Nullable ClassLoader classLoader,
                                           @Nullable Iterable<? extends Id> owners)
-    throws DatasetManagementException, IOException {
+    throws DatasetManagementException, IOException, ServiceUnavailableException {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: ditto

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java [hunk 14]
```diff
@@ -181,7 +182,7 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
                                           @Nullable Map<String, String> arguments, @Nullable ClassLoader classLoader,
                                           DatasetClassLoaderProvider classLoaderProvider,
                                           @Nullable Iterable<? extends Id> owners)
-    throws DatasetManagementException, IOException {
+    throws DatasetManagementException, IOException, ServiceUnavailableException {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: ditto

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java [hunk 15]
```diff
@@ -329,7 +279,42 @@ public Program(Application application, ProgramType type, final String id) {
...
+    public static Program from(Id.Namespace namespaceId, String appId, ProgramType type, String pgmId) {
+      return new Program(new Application(namespaceId, appId), type, pgmId);
+    }
+
+    public static Program from(String namespaceId, String appId, ProgramType type, String pgmId) {
+      return new Program(new Application(new Namespace(namespaceId), appId), type, pgmId);
+    }
+
+    public static Program fromString(String string) {
+      String[] parts = string.split(":", 2);
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: always scared of these string to object decodings. Where is this used?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java (hunk 4 above)
Comment (poornachandra) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java]: Asserting on message is always brittle. It might be enough to just assert on the exception class itself, right?

File: cdap-proto/src/test/java/co/cask/cdap/proto/codec/NamespacedIdCodecTest.java [hunk 16]
```diff
@@ -40,12 +40,12 @@
   private final Id.Namespace ns = Id.Namespace.from("ns");
   private final Id.Application app = Id.Application.from(ns, "app");
   private final Id.Program program = Id.Program.from(app, ProgramType.CUSTOM_ACTION, "action");
-  private final Id.Flow flow = Id.Flow.from(app, "flow");
```
Comment (bdmogal) [inline cdap-proto/src/test/java/co/cask/cdap/proto/codec/NamespacedIdCodecTest.java]: This pattern is consistent across this PR. So can't users ever do ``Id.Flow flow = Id.Flow.from(app, "flow")`` now? Does the type always have to be declared as Program? I think that's because ``Id.Flow`` and others don't extend ``Id.Program``. What was the reason for that change?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/ConsumerSupplier.java (hunk 3 above)
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/ConsumerSupplier.java]: `List` is not needed, only `Iterable`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java (hunk 2 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: Should we make it simple and then just remove ``Id.Workflow`` and other classes? It seems weird to me (probably because I've now gotten used to it :-) ), that ``Id.Workflow``, etc does not extend ``Id.Program``. If they are supposed to be helper classes, lets move them outside ``Id`` itself, into a helper ``Programs`` class. But even that is an overkill for saving on just a single ``ProgramType`` parameter.

File: cdap-proto/src/test/java/co/cask/cdap/proto/codec/NamespacedIdCodecTest.java (hunk 16 above)
Comment (jwang47) [inline cdap-proto/src/test/java/co/cask/cdap/proto/codec/NamespacedIdCodecTest.java]: Currently it's possible to have an `Id.Program` and `Id.Flow` that represents the same thing, but  it's better to only have `Id.Program` since it's less confusing.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java (hunk 2 above)
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: True, we can simply remove them.

**Comment by bdmogal** [comment]
@jwang47 some comments. Major concern is the helper classes, which IMO are not adding any significant value, now that they do not extend ``Id.Program``.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 15 above)
Comment (jwang47) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Oops. It can be removed now.

**Comment by bdmogal** [comment]
@jwang47, sorry, I thought about this a little more. Aren't our proto classes equivalent to public APIs? Is there a way we can do this while keeping the classes?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java [hunk 17]
```diff
@@ -67,13 +67,13 @@
...
    */
-  private final List<Id> owners;
+  private final List<? extends Id.NamespacedId> owners;
   private final Id.Run run;
   private final LineageWriter lineageWriter;
 
   @Inject
   public DefaultStreamWriter(@Assisted("run") Id.Run run,
-                             @Assisted("owners") List<Id> owners,
+                             @Assisted("owners") List<? extends Id.NamespacedId> owners,
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: Isn't this Id.Program?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java [hunk 18]
```diff
@@ -392,8 +394,11 @@ public void testBatchStatus() throws Exception {
     Assert.assertEquals(new NotFoundException(Id.Program.from("testnamespace1", "WordCountApp", ProgramType.FLOW,
                                                               "NotExist")).getMessage(),
                         returnedBody.get(0).get("error").getAsString());
-    Assert.assertEquals("'namespace:testnamespace1/application:WordCountApp/program:flows:NotExist' was not found.",
-                        returnedBody.get(0).get("error").getAsString());
+    // namespace:testnamespace1/application:WordCountApp/program:flows:NotExist
+    Assert.assertEquals(
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java]: Better define the Id as a variable and use it twice

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java [hunk 19]
```diff
@@ -392,8 +394,11 @@ public void testBatchStatus() throws Exception {
...
+      String.format("'%s' was not found.",
+                    Id.Program.from(TEST_NAMESPACE1, "WordCountApp", ProgramType.FLOW, "NotExist").toString()),
+      returnedBody.get(0).get("error").getAsString());
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java]: indentation is off

**Comment by anew** [comment]
This is a huge change, so late before release? It is not clear to me what bug it fixes?

**Comment by anew** [comment]
Actually, I could see convenience methods Id.<ProgramType>.from(...) that return an Id.Program. That would also have kept the change a lot smaller.

**Comment by anew** [comment]
If you make this huge change, why not rename the getId() method to getName() at the same time? A id is a composition of names. It is quite confusing to say that programId.getId() returns the name of the program.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java [hunk 20]
```diff
@@ -84,16 +89,18 @@ public EndpointStrategy get() {
...
+                                 @Nullable Iterable<? extends Id.NamespacedId> owners)
+    throws DatasetManagementException, ServiceUnavailableException {
 
     String query = "";
     if (owners != null) {
-      Set<String> ownerParams = Sets.newHashSet();
-      for (Id owner : owners) {
-        ownerParams.add("owner=" + owner.getIdType() + "::" + owner.getIdRep());
+      try {
+        String ownersJson = GSON.toJson(owners, new TypeToken<List<Id.NamespacedId>>() { }.getType());
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: is everything in that JSON a valid URL character?

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: It got URL encoded in the next line.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java [hunk 21]
```diff
@@ -97,8 +99,7 @@ public void list(HttpRequest request, HttpResponder responder,
    *
    * @param namespaceId namespace of the dataset instance
    * @param name name of the dataset instance
-   * @param owners a list of owners of the dataset instance, in the form @{code <type>::<id>}
-   *               (e.g. "program::namespace:default/application:PurchaseHistory/program:flow:PurchaseFlow")
+   * @param owners a UTF-8-encoded JSON list of Id owners of the dataset instance
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java]: so we would have something like ?owners=[ns1.app1.flow1,ns1.app2.svc3] in the url? Normally, to specify multiple values for a parameter in an URL, I believe, you would use repetition, as in ?owner=ns1.app1.flow1@owner=ns1.app2.svc3

Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java]: Is this a public REST API? If so, we need to keep the old syntax and deprecate it

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java [hunk 22]
```diff
@@ -415,9 +415,9 @@ private void addProgram(MDSKey.Builder keyBuilder, Id.Program program) {
   }
 
   private void addComponent(MDSKey.Builder keyBuilder, Id component) {
-    if (component instanceof Id.Flow.Flowlet) {
+    if (component instanceof Id.Flowlet) {
       keyBuilder.add(FLOWLET_MARKER)
-        .add(component.getId());
+        .add(((Id.Flowlet) component).getId());
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java]: why is the cast needed?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java [hunk 23]
```diff
@@ -180,8 +181,8 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
   public <T extends Dataset> T getDataset(Id.DatasetInstance datasetInstanceId,
                                           @Nullable Map<String, String> arguments, @Nullable ClassLoader classLoader,
                                           DatasetClassLoaderProvider classLoaderProvider,
-                                          @Nullable Iterable<? extends Id> owners)
-    throws DatasetManagementException, IOException {
+                                          @Nullable Iterable<? extends Id.NamespacedId> owners)
+    throws DatasetManagementException, IOException, ServiceUnavailableException {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: is this an unrelated change?

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 1 above)
Comment (anew) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: getId is a getter and should simply return the value of the id field. Actually it would be better to rename id to name and then you'd have getName() return "flow1", getFullName() return "namespace1.app1.flow.flow1", and toString() returns  "program:namespace1.app1.flow.flow1"

File: cdap-proto/src/test/java/co/cask/cdap/proto/codec/NamespacedIdCodecTest.java (hunk 16 above)
Comment (anew) [inline cdap-proto/src/test/java/co/cask/cdap/proto/codec/NamespacedIdCodecTest.java]: and less convenient. Also less type-safe. Id.Program could be abstract to remove the concern that you can have an Id.Program and an Id.Flow representing the same thing.

**Comment by anew** [comment]
I have done one pass. Please see my comments.

**Comment by jwang47** [comment]
@anew @bdmogal Good points, I will rework this PR so that it doesn't modify any public API classes.

**Comment by poornachandra** [comment]
@jwang47 Lets hold-off on this PR for release 3.2. We need some more discussion on this. We'll work on this for release 3.3.

**Comment by bdmogal** [comment]
@jwang47 , i think for 3.2, we should probably just fix the problem mentioned in the jira (about making sure that all id classes have the same serialization pattern) Lets do refactoring later.
